parlai/utils/torch.py

```diff
@@ -42,6 +45,37 @@ def neginf(dtype: torch.dtype) -> float:
         return -NEAR_INF
 
 
+def to_cpu(obj) -> Any:
+    """
+    Move a state-dict object to CPU.
+
+    If an obj is a dict or list, recursively iterates to place onto cpu.
+    """
+    if isinstance(obj, dict):
+        return {k: to_cpu(v) for k, v in obj.items()}
+    elif isinstance(obj, tuple):
+        return tuple(to_cpu(v) for v in obj)
+    elif isinstance(obj, list):
+        return [to_cpu(v) for v in obj]
+    elif isinstance(obj, torch.Tensor):
+        return obj.cpu()
+    else:
+        return obj
+
+
+def atomic_save(state_dict: Any, path: str) -> None:
+    """
+    Atomic save is a wrapper around torch.save, but does so in an atomic manner.
+    """
+    # atomic save is used for async writes, and it helps to yield the thread
+    time.sleep(1.0)
+    timestamp = int(time.time())
+    randint = random.randint(0, 99)
+    tmppath = path + f'.tmp.{timestamp}.{randint}'
```

Comment:
Why not use Python stdlib's `tempfile` (https://docs.python.org/3/library/tempfile.html)? :)

Reply:
I want it to be inside the NFS directory so the move is free. If I used tempfile, the file would be saved on local disk, and the rename into NFS would involve the same slow data transfer that's frighteningly preëmptable.

That said, maybe the best pattern would be
1. dump to local disk synchronously
2. launch a background thread that copies to NFS temp file, and does the rename at the end.

Reply:
Makes sense, but if we were to use a `TemporaryFile(dir=os.path.dirname(path))` we would be in the target filesystem still and avoid the teeny tiny probability of a conflict---not that that is likely at all, but just feels like the right™️ thing to do 😅

Reply:
Docs say of TemporaryFile: "It will be destroyed as soon as it is closed (including an implicit close when the object is garbage collected)."

(I don't really want a temporary file, I want a permanent file. I just want a temporary name.)

Reply:
NamedTemporaryFile looks like it has what we need tho, I'll switch to that